mmseg/models/backbones/beit.py

```diff
@@ -0,0 +1,558 @@
+
+            if 'state_dict' in checkpoint:
+                state_dict = checkpoint['state_dict']
+            else:
+                state_dict = checkpoint
+
+            all_keys = list(state_dict.keys())
+            for key in all_keys:
+                if 'relative_position_index' in key:
+                    state_dict.pop(key)
+                # In order to keep the center of pos_bias as consistent as
+                # possible after interpolation, and vice versa in the edge
+                # area, the geometric sequence interpolation method is adopted.
+                if 'relative_position_bias_table' in key:
+                    rel_pos_bias = state_dict[key]
+                    src_num_pos, num_attn_heads = rel_pos_bias.size()
+                    dst_num_pos, _ = self.state_dict()[key].size()
+                    dst_patch_shape = self.patch_shape
+                    if dst_patch_shape[0] != dst_patch_shape[1]:
+                        raise NotImplementedError()
+                    num_extra_tokens = dst_num_pos - (
+                        dst_patch_shape[0] * 2 - 1) * (
+                            dst_patch_shape[1] * 2 - 1)
+                    src_size = int((src_num_pos - num_extra_tokens)**0.5)
+                    dst_size = int((dst_num_pos - num_extra_tokens)**0.5)
+                    if src_size != dst_size:
+                        extra_tokens = rel_pos_bias[-num_extra_tokens:, :]
+                        rel_pos_bias = rel_pos_bias[:-num_extra_tokens, :]
+
+                        def geometric_progression(a, r, n):
+                            return a * (1.0 - r**n) / (1.0 - r)
+
+                        left, right = 1.01, 1.5
+                        while right - left > 1e-6:
+                            q = (left + right) / 2.0
+                            gp = geometric_progression(1, q, src_size // 2)
+                            if gp > dst_size // 2:
+                                right = q
+                            else:
+                                left = q
+
+                        dis = []
+                        cur = 1
+                        for i in range(src_size // 2):
+                            dis.append(cur)
+                            cur += q**(i + 1)
+
+                        r_ids = [-_ for _ in reversed(dis)]
+
+                        x = r_ids + [0] + dis
+                        y = r_ids + [0] + dis
+
+                        t = dst_size // 2.0
+                        dx = np.arange(-t, t + 0.1, 1.0)
+                        dy = np.arange(-t, t + 0.1, 1.0)
+
+                        all_rel_pos_bias = []
+
+                        for i in range(num_attn_heads):
+                            z = rel_pos_bias[:,
+                                             i].view(src_size,
+                                                     src_size).float().numpy()
+                            f = interpolate.interp2d(x, y, z, kind='cubic')
```

Comment:
ref: https://gist.github.com/peteflorence/a1da2c759ca1ac2b74af9a83f69ce20e#testing-for-correctness